cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -1,70 +1,164 @@
+        <*> readTMVar tmvResultLocalState
+
+  connectToLocalNodeWithVersion
+    connectInfo
+    (\ntcVersion ->
+      LocalNodeClientProtocols
+      { localChainSyncClient    = LocalChainSyncClient $ chainSyncGetCurrentTip waitResult tmvResultChainTip
+      , localStateQueryClient   = Just $ setupLocalStateQueryExpr waitResult mpoint tmvResultLocalState (runExceptT (f ntcVersion))
+      , localTxSubmissionClient = Nothing
+      }
+    )
 
-          , Net.Query.recvMsgFailure = \failure -> do
-              atomically $ putTMVar resultVar' (Just (Left failure))
+  fmap (join . first mapError) <$> atomically waitResult
+
+  where
+    chainSyncGetCurrentTip
+      :: STM b
+      -> TMVar  ChainTip
+      -> ChainSyncClient (BlockInMode mode) ChainPoint ChainTip IO ()
+    chainSyncGetCurrentTip waitDone tipVar =
+      ChainSyncClient $ pure clientStIdle
+      where
+        clientStIdle :: Net.Sync.ClientStIdle (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStIdle =
+          Net.Sync.SendMsgRequestNext clientStNext (pure clientStNext)
+
+        clientStNext :: Net.Sync.ClientStNext (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStNext = Net.Sync.ClientStNext
+          { Net.Sync.recvMsgRollForward = \_block tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
+              void $ atomically waitDone
+              pure $ Net.Sync.SendMsgDone ()
+          , Net.Sync.recvMsgRollBackward = \_point tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
               void $ atomically waitDone
-              pure $ Net.Query.SendMsgDone ()
+              pure $ Net.Sync.SendMsgDone ()
           }
-      else do
-        atomically $ putTMVar resultVar' Nothing
+
+-- | Use 'queryExpr' in a do block to construct monadic local state queries.
+setupLocalStateQueryExpr ::
+     STM x
+  -> Maybe ChainPoint
+  -> TMVar (Either Net.Query.AcquireFailure a)
+  -> LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO a
+  -> Net.Query.LocalStateQueryClient (BlockInMode mode) ChainPoint (QueryInMode mode) IO ()
+setupLocalStateQueryExpr waitDone mPointVar' resultVar' f =
+  LocalStateQueryClient . pure . Net.Query.SendMsgAcquire mPointVar' $
+    Net.Query.ClientStAcquiring
+    { Net.Query.recvMsgAcquired = runContT (runLocalStateQueryExpr f) $ \result -> do
+        atomically $ putTMVar resultVar' (Right result)
+        void $ atomically waitDone
+        pure $ Net.Query.SendMsgRelease $ pure $ Net.Query.SendMsgDone ()
+
+    , Net.Query.recvMsgFailure = \failure -> do
+        atomically $ putTMVar resultVar' (Left failure)
         void $ atomically waitDone
         pure $ Net.Query.SendMsgDone ()
+    }
+
+-- | Use 'queryExpr' in a do block to construct monadic local state queries.
+queryExpr :: Monad m => query a -> ExceptT e (LocalStateQueryExpr block point query r m) a
+queryExpr q = ExceptT $ do
```

Comment:
Why do we wrap in `ExceptT`? It looks like we never fail. Similarly in `determineEraExpr`

Reply:
This is mostly for convenience because the expression that needs to be provided must also have `ExceptT`.  To not this be the case, I need a combinator that can easily lift `Monad m => m a` to `Monad m => ExceptT e m a`.

Reply:
Okay, I was able to use `lift` to remove some `ExceptT`s.  Not sure why it was so hard the first time.